Processing
==========

Third party packages
*********************
+-------------------------------+------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Packages                      | Version                                  | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
+-------------------------------+------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `MSGFPlus`_                   | v20190628                                | performs peptide identification by scoring MS/MS spectra against peptides derived from a protein sequence database (FASTA files). Individual peptide sequences are identified(mzML file), then the set of peptide sequences is used to infer which proteins may have been present.  It reads an open data format for MS/MS identification i.e mzML files and searches again a protein database(FASTA) to outputs a .mzId file which constitutes a set of scored PSMs along with E-value estimates( i.e computes E-values of PSMs and estimates FDRs) |
+-------------------------------+------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `Mzid-To-Tsv-Converter`_      | v1.3.3                                   | to converts MS-GF+ output (.mzid) into the tsv format (.tsv).                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
+-------------------------------+------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `PeptideHitResultsProcessor`_ | v1.5.7130                                | converts tsv format (.tsv) to syn.txt file used for downstream analyses.                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |
+-------------------------------+------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `pwiz-bin-windows`_           | x86_64-vc141-release-3_0_20149_b73158966 |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
+-------------------------------+------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `MASIC`_                      | v3.0.7235                                | extracts intensity information for the identified peptides. It accurately measures peptide abundances and elution times in an LC-MS/MS analysis. It reads .Raw files from Thermo Fisher mass spectrometers and generates .SIC files i.e “selected ion chromatograms” (SICs) for each species chosen for MS/MS fragmentation.                                                                                                                                                                                                                         |
+-------------------------------+------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `sqlite-netFx-full-source`_   | 1.0.111.0                                |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
+-------------------------------+------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `msConvert`_                  |                                          | A command-line tool converting to/from various mass spectrometry data formats including multiple proprietary formats. It converts.RAW to mzML. Converting to mzML which is an open data format makes easy for academic scientists to directly manipulate MS/MS spectrum data. Open formats enable improved data sharing by allowing the data to be read by a variety of software tools without licensing restrictions                                                                                                                                |
+-------------------------------+------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `Conda`_                      | (3-clause BSD)                           |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
+-------------------------------+------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+


Data flow diagram:
------------------


Input:
------
* `.raw`, `metagenome`, `parameter files : MSGFplus & MASIC`, `contaminant_file`

Output:
-------
* `.mzml`, `mzid`, `tsv`


Credits:
--------
#. `MoTrPAC`_.

#. MS-GF+: Universal Database Search Tool for Mass Spectrometry.
        Sangtae Kim, Pavel A. Pevzner, Nat Commun. 2014 Oct 31;5:5277. doi: 10.1038/ncomms6277.

        This software is Copyright © 2012, 2013 The Regents of the University of California. All Rights Reserved.
        Permission to copy, modify, and distribute this software and its documentation for educational, research and non-profit purposes, without fee, and without a written agreement is hereby granted, provided that the above copyright notice, this paragraph and the following three paragraphs appear in all copies.

#. MASIC+: (MS/MS Automated Selected Ion Chromatogram generator) a software program for fast quantitation and flexible visualization of chromatographic profiles from detected LC-MS(/MS) features.
        Matthew E. Monroe and Shaw, Jason L and Daly, Don S and Adkins, Joshua N and Smith, Richard D
        Licensed under the 2-Clause BSD License; you may not use this file except in compliance with the License. You may obtain a copy of the License at https://opensource.org/licenses/BSD-2-Clause



Moving on
---------
Now it is time to move on to :doc:`../components/2-aggregatingAnalysis`.




.. _MoTrPAC: https://github.com/MoTrPAC/motrpac-proteomics-pnnl-prototype
.. _MzidToTSVConverter: https://github.com/PNNL-Comp-Mass-Spec/Mzid-To-Tsv-Converter

.. _MSGFPlus`:
.. _`Mzid-To-Tsv-Converter`: https://msgfplus.github.io/msgfplus/MzidToTsv.html
.. _PeptideHitResultsProcessor: https://omics.pnl.gov/software/peptide-hit-results-processor
.. _sqlite-netFx-full-source:
.. _MASIC:
.. _Conda:
.. _pwiz-bin-windows:
.. _msConvert: http://proteowizard.sourceforge.net/tools/msconvert.html